### ISSUE TITLE
[RFC]: Return extracted hidden states in the generation response

### ISSUE THREAD

**Issue by Davidnet** [vllm-project/vllm#48743]
### Motivation.

The hidden-state extraction connector currently writes its result to shared storage and returns the path to the caller. This is useful when the result is large or needs to outlive the request, but it makes even simple use cases depend on shared storage. The caller also has to read and clean up a separate file after generation completes.

For clients that want to inspect the hidden states from a single request, it would be more convenient to return the tensors with the generation result.

This RFC proposes an opt-in inline mode:

```json
{
  "kv_transfer_params": {
    "return_inline": true
  }
}
```

In this mode, the connector would skip the file write and return the extracted hidden states and their token IDs as part of the response.

Inline return is not intended to replace the existing file-based path or become the default. A hidden-state payload can be large: its size grows with the number of tokens, extracted layers, and model hidden size. The proposal trades response bandwidth and latency for a simpler request lifecycle. Callers should be able to choose which tradeoff fits their workload.

## Who wants per-request hidden states?

The main audience is interpretability tooling built on sparse autoencoders (SAEs). These workflows are inherently interactive: send one prompt, grab the residual-stream activation at a given layer, run it through a pretrained SAE, and read off which features fired. The activations are small (one hidden vector per token, often just the last token) and are consumed immediately — there is nothing to persist and no reason to route the exchange through shared storage.

Three concrete examples:

**Gemma Scope** publishes SAEs trained on the residual stream of Gemma models. The canonical usage pattern loads an SAE for a chosen layer, captures the activation at that layer for a prompt, and inspects the top-activating features

**kiji-inspector** applies the same pattern to agent tool-selection analysis: it ships SAEs with per-feature natural-language descriptions, and the core loop is "run a request, take the last-token activation, describe what fired":

```python
from kiji_inspector import SAE

sae, feature_descriptions = SAE.from_pretrained(
    repo_id="575-lab/kiji-inspector-google-gemma-4-E4B-it-home-repair",
    layer=8,
)

# Activation for the last token of the request we just served
last_token_act = hidden_state

sae.describe(last_token_act, feature_descriptions)
```

**Qwen-Scope**: Also publishes SAEs trained on the residual stream of Qwen models.

Today all of these require either running the model locally under `transformers` with forward hooks.

### User-facing behaviour

For offline generation, the returned values are CPU `torch.Tensor` objects:

```python
output = llm.generate(
    "The future of AI is",
    SamplingParams(
        max_tokens=1,
        extra_args={
            "kv_transfer_params": {
                "return_inline": True,
            }
        },
    ),
)[0]

hidden_states = output.kv_transfer_params["hidden_states"]
token_ids = output.kv_transfer_params["token_ids"]
```

For the Python HTTP frontend, each tensor is encoded as a JSON object with `dtype`, `shape`, and base64-encoded `data` fields. The hidden-state connector provides `decode_base64_tensor()` for reconstructing the tensor on the client. The encoding is applied at the serving boundary so the engine can continue to work with tensors rather than a transport-specific representation.

The response also includes `"hidden_states_inline": true`, which lets callers distinguish this mode from the existing `hidden_states_path` result.

### Connector contract

There is a small timing problem behind this feature: the model can produce its last token before the connector has finished copying the extracted tensors to CPU. If the engine sends the final output immediately, the payload arrives too late to be included.

To handle that, we propose `take_finished_sending_payloads()` to the V1 connector interface. It returns payloads in the following form:

```python
{
    request_id: {
        "hidden_states": hidden_states,
        "token_ids": token_ids,
    }
}
```

The tensors must be on CPU, and a connector must make a request's payload available no later than the step where it reports that request in `finished_sending`. `MultiConnector` and the worker output aggregator merge the named tensors from multiple connectors or workers.

Connectors that do not return data inline do not need to change their behavior; the new method defaults to `None`.

### Holding the final output

When the hidden-state connector sees `return_inline`, its `request_finished()` result asks the scheduler to defer the final output. The scheduler holds that `EngineCoreOutput` until the corresponding `finished_sending` notification arrives, attaches the CPU tensors, and then emits the output normally.

The defer marker is internal connector metadata something like a `KV_DEFER_OUTPUT_KEY` global and is removed before anything is returned to the caller. Payloads have a separate `kv_transfer_tensors` field on `EngineCoreOutput`; the output processor merges that field into the public `RequestOutput.kv_transfer_params` dictionary. This keeps tensors typed while they cross the engine boundary instead of hiding them inside an untyped metadata dictionary.

### Python/Rust wire compatibility

`EngineCoreOutput` is encoded positionally, so adding a field in the middle of the struct would shift every field after it. `kv_transfer_tensors` is therefore a trailing, optional field. Python omits it when unused, while the Rust engine-core client accepts it as an opaque value. Compatibility fixtures cover both the ordinary output and an output containing an inline tensor payload.

The Rust frontend cannot currently reconstruct the tensor frames carried by the Python engine protocol. For this first version, it rejects requests with `return_inline` instead of accepting them and losing the payload. Supporting inline tensors in the Rust frontend can be considered separately once there is a consumer for it.

### Initial limitations

The first version only supports requests for which the existing API has one unambiguous place to return the payload: a non-streaming request with one prompt and one generated sequence.

The Python serving layer therefore rejects inline return with:

- streaming;
- more than one prompt;
- `n > 1`;
- beam search; or
- builtin tool execution in the Responses API.

The last three cases fan out into multiple child requests or engine turns, but the public response currently has only one `kv_transfer_params` value. Choosing one child's or one turn's tensors would silently discard the others. The engine also checks the `n > 1` case so offline and direct `AsyncLLM` callers do not bypass the serving validation.

These restrictions are about the current response contract, not a claim that streaming or multi-output support is impossible. If those use cases become important, we should first decide whether payloads belong to each choice, each turn, or a final aggregate.

### Alternatives considered

1. Keep file-based return as the only option. This avoids response overhead,
   but retains the shared-storage and cleanup requirements for every caller.
2. Put encoded tensors directly in `kv_transfer_params` inside EngineCore.
   That couples the engine to HTTP serialization and makes the Python/Rust
   protocol harder to type safely.
3. Add aggregation for sampling, beam search, and tool calls now. There is no
   established API shape for those payloads yet, so rejecting them is safer
   than introducing a contract we may need to undo.

### Compatibility

There is no behaviour change unless `return_inline` is set. The new `EngineCoreOutput` field is optional and trailing, and connectors inherit a no-op implementation of the new hook.

### Testing

Coverage should include:

- connector extraction and inline payload collection;
- deferred scheduler output, including payloads that arrive before or with
  `finished_sending`;
- multi-worker and `MultiConnector` payload merging;
- validation for unsupported request shapes;
- offline and OpenAI-compatible response behaviour; and
- Python/Rust MessagePack compatibility and Rust-side rejection.

### Feedback Period.

Two weeks.

### Any Other Things.

The main questions I would like feedback on are:

- Is `kv_transfer_params` the right public home for connector-produced
  tensors, or should this be a separate output field?
- Is a generic connector payload hook useful, or should the mechanism remain
  specific to hidden-state extraction?
- For future multi-output support, should payloads be associated with choices,
  engine turns, or only the final aggregated response?

I would appreciate some feedback on this implementation:
https://github.com/Davidnet/vllm/commit/b6455d43be849d4850bed6ecfb834489ba9f0a08

I used AI assistance to ensure the coherence and grammatical accuracy of the RFC. I also utilized it for the implementation PR proposal.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Davidnet** [comment]
A few benchmarks on using the models:

- Nemotron
- Gemma models
- Qwen3.6

is tracked in:
https://github.com/575-lab/residual-stream-model-comparisons

**Comment by SorenDreano** [comment]
I’d also be interested in this feature.

Being able to return hidden states directly through the openai API (rather than the server writing on its disk and using SSHFS/NFS to get the safetensors) would make vLLM much easier to use for representation extraction, probing, and some forms of embeddings.

In my case, even support for returning just the final-layer hidden state would already be extremely useful.

**Comment by qgallouedec** [comment]
+1 for the narrow version: final-layer hidden state in the generation response. (No layer selection, no windows.)

On-policy distillation in TRL scores student completions against a teacher served by vLLM, teacher-forced with `prompt_logprobs=k`. The loss needs the teacher's full-vocab distribution. Top-k is not it, capped at `max_logprobs` (20), so we narrow the support to two candidates and document that as a limitation: https://github.com/huggingface/trl/blob/main/trl/experimental/async_distillation/async_distillation_config.py#L107-L112

The final hidden state is the cheap exact version: 4096 bf16 values per token on Qwen3-8B instead of a 151936-wide distribution, reconstructed trainer-side against the teacher's `lm_head`.

`extract_hidden_states` plus the example KV connector does not cover it, it writes safetensors to the server's disk and our teacher is remote.
